yb-voyager/cmd/common.go

```diff
@@ -526,3 +528,75 @@ func getDefaultPGSchema(schema string) (string, bool) {
 		return "", true
 	}
 }
+
+func CleanupChildProcesses() {
+	myPid := os.Getpid()
+	processes, err := ps.Processes()
+	if err != nil {
+		log.Errorf("failed to read processes: %v", err)
+	}
+	childProcesses := lo.Filter(processes, func(process ps.Process, _ int) bool {
+		return process.PPid() == myPid
+	})
+	lo.ForEach(childProcesses, func(process ps.Process, _ int) {
+		pid := process.Pid()
+		log.Infof("shutting down child pid %d", pid)
+		err := ShutdownProcess(pid, 60)
+		if err != nil {
+			log.Errorf("shut down child pid %d : %v", pid, err)
+		} else {
+			log.Infof("shut down child pid %d", pid)
+		}
+	})
+}
+
+// this function wait for process to exit after signalling it to stop
+func ShutdownProcess(pid int, forceShutdownAfterSeconds int) error {
+	if forceShutdownAfterSeconds > 0 {
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		go func(ctx context.Context) {
+			time.Sleep(time.Duration(forceShutdownAfterSeconds) * time.Second)
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				log.Infof("force shutting down child pid %d", pid)
+				signalProcess(pid, syscall.SIGKILL)
+			}
+		}(ctx)
+	}
```

Comment:
You can avoid all this complexity. The `waitForProcessToExit()` is polling for process exit after every 2 seconds. The function can keep track whether it has been longer than `forceShutdownAfterSeconds` -- if yes, kill the process.

Reply:
Indeed, thanks! Guess I got a little carried away with goroutines and contexts 😅 
Made the changes.